downstairs/src/region.rs

```diff
@@ -1426,18 +1455,25 @@ impl Extent {
         cdt::extent__flush__sqlite__insert__start!(|| {
             (job_id, self.number, n_dirty_blocks)
         });
-        inner.truncate_encryption_contexts_and_hashes(
+
+        // We put all of our metadb updates into a single transaction to
+        // assure that we have a single sync.
+        let tx = inner.metadb.unchecked_transaction()?;
+
+        inner.truncate_encryption_contexts_and_hashes_with_tx(
             extent_block_indexes_and_hashes,
+            &tx,
         )?;
+
         cdt::extent__flush__sqlite__insert__done!(|| {
             (job_id, self.number, n_dirty_blocks)
         });
 
-        // Reset the file's seek offset to 0, and set the flush number and gen
-        // number
-        inner.file.seek(SeekFrom::Start(0))?;
-
         inner.set_flush_number(new_flush, new_gen)?;
+        tx.commit()?;
```

Comment:
if this works, i don't understand why. this code is not passing the transaction into `set_flush_number()` which means the sqlite queries inside `set_flush_number()` aren't using the transaction, right? what *are* they doing? I have the same question for setting the dirty flag in `write()` further up the file.

Reply:
> Though, given I was reading that `.transaction()` returns an `&mut` to make the borrow checker statically avoid nested transactions, maybe this makes sense. Is it just the case that `Transaction` providing its own functions for getting prepared statements is to work within the limits of rust's type system, and that it doesn't actually reflect the way sqlite is used at the C-api level?

Yes, this is right. The SQLite C API doesn't have a notion of transactions at all - it just has a connection, and opening/committing/rolling back a transaction all happen via SQL statements executed on that connection. This is most obvious from rusqlite's point of view in [`Transaction::new_unchecked()`](https://docs.rs/rusqlite/0.29.0/src/rusqlite/transaction.rs.html#121-129): it executes a `BEGIN ...` and then just wraps the connection, which we are now tracking at the type system level as being in a transaction. (This can be trivially foiled by a program that executes its own `BEGIN/COMMIT/ROLLBACK` SQL that rusqlite doesn't know about.)

Reply:
An idle thought looking at `Transaction::new_unchecked()` again (and remembering #436 go by): it calls [`conn.execute_batch("BEGIN ...")`](https://docs.rs/rusqlite/0.29.0/src/rusqlite/transaction.rs.html#126) which internally calls [`conn.prepare(...)`](https://docs.rs/rusqlite/0.29.0/src/rusqlite/lib.rs.html#501), _not_ `conn.prepare_cached()`. So every transaction opened (and committed, which uses the same mechanism) incurs a new statement preparation. I don't know whether that has a material perf impact but wanted to mention it.